https/tls_config.go

```diff
@@ -0,0 +1,109 @@
+// Package https allows the implementation of tls
+package https
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/prometheus/common/log"
+	"gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	TLSCertPath string    `yaml:"tlsCertPath"`
+	TLSKeyPath  string    `yaml:"tlsKeyPath"`
+	TLSConfig   TLSStruct `yaml:"tlsConfig"`
+}
+
+type TLSStruct struct {
+	RootCAs            string `yaml:"rootCAs"`
+	ServerName         string `yaml:"serverName"`
+	ClientAuth         string `yaml:"clientAuth"`
+	ClientCAs          string `yaml:"clientCAs"`
+	InsecureSkipVerify bool   `yaml:"insecureSkipVerify"`
+}
+
+func GetTLSConfig(configPath string) *tls.Config {
+	config, err := loadConfigFromYaml(configPath)
+	if err != nil {
+		log.Error("Config failed to load from Yaml", err)
+	}
+	tlsc, err := LoadTLSConfig(config)
+	if err != nil {
+		log.Error("Failed to convert Config to tls.Config", err)
+	}
+	return tlsc
+}
+
+func loadConfigFromYaml(fileName string) (*Config, error) {
+	content, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return nil, err
+	}
+	c := &Config{}
+	err = yaml.Unmarshal(content, c)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func LoadTLSConfig(c *Config) (*tls.Config, error) {
+	cfg := &tls.Config{}
+	if len(c.TLSCertPath) > 0 {
+		cfg.GetCertificate = func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			cert, err := tls.LoadX509KeyPair(c.TLSCertPath, c.TLSKeyPath)
+			if err != nil {
+				return nil, err
+			}
+			return &cert, nil
+		}
+	}
+	cfg.ServerName = c.TLSConfig.ServerName
+
+	cfg.InsecureSkipVerify = c.TLSConfig.InsecureSkipVerify
+
+	if len(c.TLSConfig.RootCAs) > 0 {
+		rootCertPool := x509.NewCertPool()
+		rootCAFile, err := ioutil.ReadFile(c.TLSConfig.RootCAs)
+		if err != nil {
+			return cfg, err
+		}
+		rootCertPool.AppendCertsFromPEM(rootCAFile)
+		cfg.RootCAs = rootCertPool
+	}
+	if len(c.TLSConfig.ClientCAs) > 0 {
+		clientCAPool := x509.NewCertPool()
+		clientCAFile, err := ioutil.ReadFile(c.TLSConfig.ClientCAs)
+		if err != nil {
+			return cfg, err
+		}
+		clientCAPool.AppendCertsFromPEM(clientCAFile)
+		cfg.ClientCAs = clientCAPool
+	}
+	if len(c.TLSConfig.ClientAuth) > 0 {
+		switch s := (c.TLSConfig.ClientAuth); s {
+		case "RequestClientCert":
+			cfg.ClientAuth = tls.RequestClientCert
+		case "RequireClientCert":
+			cfg.ClientAuth = tls.RequireAnyClientCert
+		case "VerifyClientCertIfGiven":
+			cfg.ClientAuth = tls.VerifyClientCertIfGiven
+		case "RequireAndVerifyClientCert":
+			cfg.ClientAuth = tls.RequireAndVerifyClientCert
+		default:
+			cfg.ClientAuth = tls.NoClientCert
+		}
+	}
+	return cfg, nil
+}
+
+func Listen(server *http.Server) error {
+	if server.TLSConfig != nil {
```

Comment:
I am guessing if someone is using a package `https` they definitely want to `ListenAndServeTLS` or an error otherwise. What do you think?

Reply:
Or document it via a function comment?

Reply:
Document it I'd say, as https is optional and it'd be better not to have to duplicate that code everywhere.